### ISSUE TITLE
svg2svg with output does not return the original

### ISSUE THREAD

**Issue by piscvau** [Kozea/CairoSVG#267]
Hello
Svg2svg does not seem to return a file identical to the input but an empty file with a clone;
alos the byte returned when output = None seems incorrect.
Input file is a very simple file from inkscape. whose content is here: 

```xml
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<!-- Created with Inkscape (http://www.inkscape.org/) -->

<svg
   xmlns:dc="http://purl.org/dc/elements/1.1/"
   xmlns:cc="http://creativecommons.org/ns#"
   xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
   xmlns:svg="http://www.w3.org/2000/svg"
   xmlns="http://www.w3.org/2000/svg"
   xmlns:sodipodi="http://sodipodi.sourceforge.net/DTD/sodipodi-0.dtd"
   xmlns:inkscape="http://www.inkscape.org/namespaces/inkscape"
   width="210mm"
   height="297mm"
   viewBox="0 0 210 297"
   version="1.1"
   id="svg8"
   inkscape:version="0.92.5 (0.92.5+68)"
   sodipodi:docname="rectinkscape.svg"
   enable-background="new">
  <defs
     id="defs2">
    <filter
       inkscape:collect="always"
       style="color-interpolation-filters:sRGB"
       id="filter833">
      <feBlend
         inkscape:collect="always"
         mode="lighten"
         in2="BackgroundImage"
         id="feBlend835" />
    </filter>
  </defs>
  <sodipodi:namedview
     id="base"
     pagecolor="#ffffff"
     bordercolor="#666666"
     borderopacity="1.0"
     inkscape:pageopacity="0.0"
     inkscape:pageshadow="2"
     inkscape:zoom="0.86704545"
     inkscape:cx="90.275229"
     inkscape:cy="548.57143"
     inkscape:document-units="mm"
     inkscape:current-layer="layer1"
     showgrid="false"
     inkscape:window-width="1600"
     inkscape:window-height="841"
     inkscape:window-x="0"
     inkscape:window-y="25"
     inkscape:window-maximized="1" />
  <metadata
     id="metadata5">
    <rdf:RDF>
      <cc:Work
         rdf:about="">
        <dc:format>image/svg+xml</dc:format>
        <dc:type
           rdf:resource="http://purl.org/dc/dcmitype/StillImage" />
        <dc:title></dc:title>
      </cc:Work>
    </rdf:RDF>
  </metadata>
  <g
     inkscape:label="Layer 1"
     inkscape:groupmode="layer"
     id="layer1"
     style="filter:url(#filter833);opacity:0.98">
    <rect
       style="opacity:0.97000002;vector-effect:none;fill:none;fill-opacity:1;fill-rule:nonzero;stroke:#130303;stroke-width:0.26458333;stroke-linecap:square;stroke-linejoin:round;stroke-miterlimit:4;stroke-dasharray:none;stroke-dashoffset:0;stroke-opacity:1"
       id="rect815"
       width="46.993885"
       height="31.125818"
       x="31.73613"
       y="87.053299" />
  </g>
</svg>
```

**Comment by liZe** [comment]
`svg2svg`’s goal is not to generate exactly the same file: it parses the original SVG file, takes what in understands, generates a Cairo surface with this information.

The generated SVG is not equivalent to the original one because there’s a filter that’s not handled by CairoSVG (see #23). This filter (lighten) is supported by Cairo and could probably be supported by CairoSVG too.

**Comment by piscvau** [comment]
The problem is not the filter. the output of the svg2svg is an image. Is this what is intended?
here is the output file which I shortened:
`<?xml version="1.0" encoding="UTF-8"?>
<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="595.275591pt" height="841.889764pt" viewBox="0 0 595.275591 841.889764" version="1.1">
<defs>
<image id="image8" width="2484" height="3509" xlink:href="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAACbQAAA21CAYAAADsC4rSAAAABmJLR0QA/wD/AP+gvaeTAAAgAElEQVR4nOzBAQEAAACAkP6v7ggKAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA...........
AACAlQABFS3AiOvUSwAAAABJRU5ErkJggg=="/>
</defs>
<g id="surface1">
<use xlink:href="#image8" transform="matrix(0.24,0,0,0.24,0,0)"/>
</g>
</svg>`

**Comment by liZe** [comment]
Oh.

Contrary to what I said, the "lighten" filter is supported by CairoSVG.

You get an image because of Cairo limitations. Applying the filter puts the Cairo surface in such a state that it doesn’t know how to translate it back into a vector format.

If you remove the filter, you get a "normal" SVG file.

**Comment by piscvau** [comment]
thanks it solves the problem. Another question. after calling svg2svg we get a vector-based single page cairo surface. Is there a way to retrieve what has been drawn in ths surface (paths .....). I tried  creating a recording surface in the __init__ but to does not seem to give me this possibility. Or should I modify the draw function so that it constructs a list of what it draws?

**Comment by liZe** [comment]
> Is there a way to retrieve what has been drawn in ths surface (paths .....).

It’s possible, but it really depends on what you really want to do from the paths you get. What’s your final goal?

**Comment by piscvau** [comment]
The final goal is to take an SVG file as input and find in the file the closed paths or shapes to make them knitting patterns for a knitting machine.These patterns are One color, no stroke. After getting the patterns, I will need to fill them, do some calculations, scale them and make them in pixels (2 colors only). One way I found  is to redefine the context methods which draw(shapes and paths)  and to create a list of these calls with their parameters. I tried it and it seems to work though it is not very elegant. If you have any better suggestion!...In short I want to use Cairo SVG to decode the file, take care of all clips, coordinates, etc... but not care about the colors gradients etc.. and give me the shpaes and paths..
